### PR TITLE
Joolian studies

### PR DESCRIPTION
**Pull request by eggrobin**
A better version of `ksp_system_test`, geared towards plotting and logging rather than shoehorning an expectation.
This tool should allow us to gain a better understanding of the magic number that is the timestep, since we can use it to vaguely estimate the global backward error.

### CODE REVIEW

File: mathematica/retrobop_dynamical_stability.cpp [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+template<typename Integrator>
+std::list<not_null<std::unique_ptr<Ephemeris<Barycentric>>>>
+MakePerturbedEphemerides(int const count,
+                         Integrator const& integrator,
+                         Time const& step) {
+  std::mt19937_64 generator;
+  std::list<not_null<std::unique_ptr<Ephemeris<Barycentric>>>> result;
+  for (int i = 0; i < count; ++i) {
+    auto system = ReadSystem(system_file);
+    for (auto const celestial : jool_system) {
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: I have no idea what the types are.  Please don't use `auto` here and on the previous line.

File: mathematica/retrobop_dynamical_stability.cpp [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+  return result;
+}
+
+void FillPositions(
+    Ephemeris<Barycentric> const& ephemeris,
+    Instant const& initial_time,
+    Time const& duration,
+    std::vector<std::vector<Vector<double, Barycentric>>>& container) {
+  for (int n = 0; n * step < duration; ++n) {
+    Instant const t = initial_time + n * step;
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: Make `n * step` the loop variable, not `n`.  We are not writing FORTRAN this week.

File: mathematica/retrobop_dynamical_stability.cpp [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+    result.emplace_back(MakeEphemeris(std::move(system), integrator, step));
+  }
+  return result;
+}
+
+void FillPositions(
+    Ephemeris<Barycentric> const& ephemeris,
+    Instant const& initial_time,
+    Time const& duration,
+    std::vector<std::vector<Vector<double, Barycentric>>>& container) {
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: `container` is obscure.  Yes, I know that a `vector` is a container.  Please pick a better name.  In particular, I'm not sure why it contains `double`.

File: mathematica/retrobop_dynamical_stability.cpp [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+}
+
+void ProduceCenturyPlots(Ephemeris<Barycentric>& ephemeris) {
+  for (int i = 1; i < (a_century_hence - ksp_epoch) / JulianYear; ++i) {
+    LOG(INFO) << "year " << i;
+    ephemeris.Prolong(ksp_epoch + i * JulianYear);
+  }
+  ephemeris.Prolong(a_century_hence);
+
+  std::map<Celestial, std::vector<double>> extremal_separations_in_m;
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: Why did we stop using quantities here and in the code that follows?

File: mathematica/retrobop_dynamical_stability.cpp [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+
+  std::map<Celestial, double> record_separation_in_m;
+
+  for (auto const moon : jool_moons) {
+    last_separation_changes.emplace(moon, Sign(+1));
+  }
+  for (int n = 0; t < a_century_hence; ++n, t = ksp_epoch + n * Hour) {
+    auto const jool_position = EvaluatePosition(ephemeris, Jool, t);
+
+    for (auto const moon : jool_moons) {
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: Explicit type.

File: mathematica/retrobop_dynamical_stability.cpp [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+
+  std::vector<double> tylo_bop_separations_in_m;
+  std::vector<double> pol_bop_separations_in_m;
+
+  std::map<Celestial, double> record_separation_in_m;
+
+  for (auto const moon : jool_moons) {
+    last_separation_changes.emplace(moon, Sign(+1));
+  }
+  for (int n = 0; t < a_century_hence; ++n, t = ksp_epoch + n * Hour) {
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: Same comment as above regarding `n`.

File: mathematica/retrobop_dynamical_stability.cpp [hunk 7]
```diff
@@ -0,0 +1,668 @@
...
+    auto const jool_position = EvaluatePosition(ephemeris, Jool, t);
+
+    for (auto const moon : jool_moons) {
+      Length const separation =
+          (jool_position - EvaluatePosition(ephemeris, moon, t)).Norm();
+      Sign const separation_change = Sign(separation - last_separations[moon]);
+      if (separation_change != last_separation_changes.at(moon)) {
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: Why `at`?

File: mathematica/retrobop_dynamical_stability.cpp [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+      bop_jacobi_nodes_in_degrees.emplace_back(
+          bop_jacobi_elements.longitude_of_ascending_node / Degree);
+      bop_jacobi_arguments_of_periapsis_in_degrees.emplace_back(
+          bop_jacobi_elements.argument_of_periapsis / Degree);
+    }
+  }
+
+  std::ofstream file;
+  file.open("retrobop_century.generated.wl");
+  file << mathematica::Assign("laytheTimes", times_in_s[Laythe]);
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: I would prefer to keep the strongly-typed quantities up to this point (that would avoid the ugly multiplications/divisions by `Metre` and friends) and have a small function to remove typing here:
```
template<typename Q>
std::vector<double> RemoveTheTypesDamnit(std::vector<Q> const& q) {
  for ... {
    result.push_back(q[i] / SIUnit<Q>());
  }
}
```

File: mathematica/retrobop_dynamical_stability.cpp [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+
+void ComputeHighestMoonError(Ephemeris<Barycentric> const& left,
+                             Ephemeris<Barycentric> const& right,
+                             Instant const& t,
+                             Length& error,
+                             Celestial& most_erroneous_moon) {
+  error = Length{};
+  auto const left_barycentre = JoolSystemBarycentre(left, t).position();
+  auto const right_barycentre = JoolSystemBarycentre(right, t).position();
+  for (auto const moon : jool_moons) {
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: Explicit type.

File: mathematica/retrobop_dynamical_stability.cpp [hunk 10]
```diff
@@ -0,0 +1,668 @@
...
+      ReadSystem(system_file),
+      integrators::QuinlanTremaine1990Order12<Position<Barycentric>>(),
+      step);
+
+  for (int i = 1; i <= 5; ++i) {
+    ephemeris->Prolong(ksp_epoch + i * JulianYear);
+    LOG(INFO) << "Prolonged to year " << i;
+  }
+
+  std::vector<std::vector<Vector<double, Barycentric>>>
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: Same comment about a better way to drop typed quantities.

File: mathematica/retrobop_dynamical_stability.cpp [hunk 11]
```diff
@@ -0,0 +1,668 @@
...
+    Bundle bundle{7};
+    if (reference_ephemeris != nullptr) {
+      bundle.Add([&reference_ephemeris = *reference_ephemeris, t ]() {
+        reference_ephemeris.Prolong(t);
+        reference_ephemeris.ForgetBefore(t);
+        return Status::OK;
+      });
+    }
+    if (refined_ephemeris != nullptr) {
+      bundle.Add([&refined_ephemeris = *refined_ephemeris, t ]() {
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: Remove space before square bracket.

File: mathematica/retrobop_dynamical_stability.cpp [hunk 12]
```diff
@@ -0,0 +1,668 @@
...
+  }
+  int total_breakdowns = 0;
+  // If the error between an integration at step and one at step/2 exceeds this
+  // over a year, we assume that things have happened that our integrator cannot
+  // handle, probably close encounters.
+  // TODO(egg): This is a very lousy substitute for a proper estimation of the
+  // local forward error.  We probably want to have a way to actually estimate
+  // the local error (on every step), and perhaps even the local backward error
+  // (though that may be costly if done naïvely).
+  Length const yearly_allowed_numerical_error = 1000 * Metre;
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: `Kilo(Metre)`

File: mathematica/retrobop_dynamical_stability.cpp [hunk 13]
```diff
@@ -0,0 +1,668 @@
...
+
+  bool log_radius = true;
+  // Errors below this are invisible on plots.
+  Length const visible_threshold = 1e6 * Metre;
+  // Errors above this mean we are pretty much completely out of phase.
+  Length const chaotic_threshold = 1e8 * Metre;
+
+  for (int year = 1;; ++year) {
+    Instant const t = ksp_epoch + year * JulianYear;
+    Bundle bundle{7};
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: Comment the 7.  Actually make it a constant, it shows up at several places.

File: mathematica/retrobop_dynamical_stability.cpp [hunk 14]
```diff
@@ -0,0 +1,668 @@
...
+      if (numerically_unsound[it->get()]) {
+        goto erase_and_skip_to_next_perturbed_ephemeris;
+      } else {
+        Ephemeris<Barycentric> const& ephemeris = **it;
+        auto const jool_barycentre =
+            JoolSystemBarycentre(ephemeris, t).position();
+        for (auto const moon : jool_moons) {
+          Length const distance =
+              (EvaluatePosition(ephemeris, moon, t) - jool_barycentre).Norm();
+          if (distance > 3e8 * Metre) {
```
Comment (pleroy) [inline mathematica/retrobop_dynamical_stability.cpp]: This kind of threshold (there are others in this code) should be made constants at the beginning of the file.

File: mathematica/retrobop_dynamical_stability.cpp (hunk 2 above)
Comment (eggrobin) [inline mathematica/retrobop_dynamical_stability.cpp]: incrementing by `step` `n` times isn't the same thing as adding `n * step`...

File: mathematica/retrobop_dynamical_stability.cpp (hunk 7 above)
Comment (eggrobin) [inline mathematica/retrobop_dynamical_stability.cpp]: No default constructor for `Sign`.

**Pull request merged**